lib/elixir/lib/map_set.ex

```diff
@@ -38,7 +38,9 @@ defmodule MapSet do
 
   @type value :: term
 
-  @opaque t(value) :: %__MODULE__{map: %{optional(value) => []}}
+  @opaque internal_data :: %{optional(value) => []}
+
+  @type t(internal_data) :: %__MODULE__{map: internal_data}
```

Comment:
That's much better. But we also need to hide the key. So I think it would be this:

```elixir
@typep internal_key :: :map 
@typep internal_value(value) :: %{optional(value) => []}
@type t(value) :: %__MODULE__{required(internal_key) => interval_value(value)}
```

Reply:
Actually, this looks so complex that I am thinking about doing what @michalmuskala suggests and just not support it. :(

Reply:
Oh sure that makes sense @josevalim, I updated with your suggestion(s), but understand if we want to take a different direction. Look forward to hearing what you decide.

Reply:
Thank you @treble37! Your solution is correct :heart: but it indeed looks very complex for reasons beyond our (yours and mine) control. We will discuss this a bit and follow up accordingly.

Reply:
👍 thank you